src/utils/rpc_selector.rs

```diff
@@ -14,30 +19,68 @@ pub enum RpcSelectorError {
     NoProviders,
     #[error("Client initialization failed: {0}")]
     ClientInitializationError(String),
+    #[error("Weighted index error: {0}")]
+    WeightedIndexError(String),
+    #[error("All available providers have failed")]
+    AllProvidersFailed,
 }
 
-/// Creates a weighted distribution for selecting RPC endpoints based on their weights.
-///
-/// # Arguments
-/// * `configs` - A slice of RPC configurations with weights
-///
-/// # Returns
-/// * `Option<Arc<WeightedIndex<u8>>>` - A weighted distribution if configs have different weights, None otherwise
-pub fn create_weights_distribution(configs: &[RpcConfig]) -> Option<Arc<WeightedIndex<u8>>> {
-    if configs.len() <= 1 {
-        return None;
+// Provider health tracking struct
+#[derive(Debug)]
+struct ProviderHealth {
+    // Set of indices that are currently marked as failed
+    failed_providers: HashSet<usize>,
+    // The time when the failed providers should be automatically reset
+    auto_reset_time: Option<Instant>,
+    // The duration after which failed providers should be automatically reset
+    reset_duration: Duration,
```

Comment:
What's the difference between these two?

Reply:
> Yeah, if the provider fails to complete a request multiple times, we mark it as failed and disable it. However, it will be available again after 5 minutes because of this logic.

Are we updating some global store to disable it?
We are creating instances on eery requests so I am not sure these updates are persisted